src/virtio/net.c

```diff
@@ -0,0 +1,385 @@
+
+        memcpy(dest_buf + written, (void *)desc->addr + skipping, writing);
+
+        skip_remaining -= skipping;
+        written += writing;
+        dest_remaining -= writing;
+
+        if (desc->flags & VIRTQ_DESC_F_NEXT) {
+            desc = &virtq->desc[desc->next];
+        } else {
+            break;
+        }
+    }
+
+    sddf_buffer.len = written;
+    error = net_enqueue_active(&state->tx, sddf_buffer);
+    /* This cannot fail as we check above */
+    assert(!error);
+
+    virtq_enqueue_used(virtq, desc_head, written);
+    *respond_to_client = true;
+    *notify_tx_server = true;
+    return;
+
+fail:
+    virtq_enqueue_used(virtq, desc_head, 0);
+    *respond_to_client = true;
+}
+
+static int virtio_net_queue_notify(struct virtio_device *dev)
+{
+    struct virtio_net_device *state = device_state(dev);
+
+    if (dev->data.QueueSel != VIRTIO_NET_TX_VIRTQ) {
+        LOG_NET_ERR("Invalid queue\n");
+        return 0;
+    }
+
+    virtio_queue_handler_t *vq = &dev->vqs[VIRTIO_NET_TX_VIRTQ];
+    struct virtq *virtq = &vq->virtq;
+
+    uint16_t guest_idx = virtq->avail->idx;
+    uint16_t idx = vq->last_idx;
+
+    bool notify_tx_server = false;
+    bool respond_to_client = false;
+
+    for (; idx != guest_idx; idx++) {
+        uint16_t desc_head = virtq->avail->ring[idx % virtq->num];
+        handle_tx_msg(dev, virtq, desc_head, &notify_tx_server, &respond_to_client);
+    }
+
+    vq->last_idx = idx;
+
+    if (notify_tx_server && net_require_signal_active(&state->tx)) {
+        net_cancel_signal_active(&state->tx);
+        if (!have_signal) {
+            microkit_notify_delayed(state->tx_ch);
+        } else if (signal_cap != BASE_OUTPUT_NOTIFICATION_CAP + state->tx_ch) {
+            microkit_notify(state->tx_ch);
+        }
+    }
+    if (respond_to_client) {
+        virtio_net_respond(dev);
+    }
+
+    return 1;
+}
+
+static uint32_t copy_rx(struct virtq *virtq,
+                        uint16_t *curr_desc_head,
+                        uint32_t *desc_copied,
+                        const void *buf, uint32_t size)
+{
+    uint32_t copied = 0;
+    do {
+        uint32_t copying = MIN(size - copied, virtq->desc[*curr_desc_head].len - *desc_copied);
+
+        memcpy((void *)virtq->desc[*curr_desc_head].addr + *desc_copied, buf + copied, copying);
+
+        copied += copying;
+        *desc_copied += copying;
+
+        if (*desc_copied == virtq->desc[*curr_desc_head].len) {
+            if (!(virtq->desc[*curr_desc_head].flags & VIRTQ_DESC_F_NEXT)) {
+                break;
+            }
+            *curr_desc_head = virtq->desc[*curr_desc_head].next;
+            *desc_copied = 0;
+        }
+    } while (copied < size);
+
+    return copied;
+}
+
+static void handle_rx_buffer(struct virtio_device *dev,
+                             uint64_t buf_offset, uint32_t size)
+{
+    struct virtio_net_device *state = device_state(dev);
+
+    if (!dev->vqs[VIRTIO_NET_RX_VIRTQ].ready) {
```

Comment:
According to the virtIO spec (3.3.1 of v1.2) we should probably be waiting until the driver sets the `DRIVER_OK` bit. It's probably unlikely to make a difference in reality but I think since the spec outlines the initialisation order of a device/driver, we should follow it.

Reply:
Done https://github.com/au-ts/libvmm/pull/95/commits/03091e3c2d99b3c0554d718033ce88fbbae4672b